[Core] Implement Fish Ranking Contest

paladindamarus opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Implement Fish Ranking contest in Selbina
- Adds Pelican Ring item and effect
- Adds database tables for contest and player fishing history (history to be implemented separately)
- Includes `!addfish` command to create fish items with size data in exdata
- Modifies packet guard to allow 0x04b packet with player in CS substate
- Updates packet handler for 0x04b (fish ranking data shares packet type with server message request)
- Adds entry to `xi.settings.main` to auto-run fishing contest

## Steps to test these changes

- Once added, server instance that loads Selbina will auto-initiate the contest
- If no contest exists, it will create one
- Speak to Chenon in Selbina to interact
- Use `!addfish` to create appropriate fish type with size/length data
- Trade fish to Chenon during appropriate stage of contest

## To-Do

- Pelican ring effect should stack to maximum 2, but issues with effect overwrite prevents this.  Require update to core to adjust?
- Create commands for manual control (note: there are menus built in to the client for GM/Staff control of contest, but it is unknown if this is used or put on auto-pilot by retail servers)

paladindamarus committed:
[Core] Implement Fish Ranking contest in Selbina

Xaver-DaRed commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+    { id = 4480, name = 'gugru_tuna'          , realname = 'Gugru Tuna'      },
+    { id = 5120, name = 'titanic_sawfish'     , realname = 'Titanic Sawfish' },
+    { id = 5127, name = 'gugrusaurus'         , realname = 'Gugrusaurus'     },
+    { id = 5129, name = 'lik'                 , realname = 'Lik'             },
```
This need xi.item enums

Xaver-DaRed commented on scripts/items/pelican_ring.lua:
```diff
@@ -0,0 +1,49 @@
+itemObject.onEffectGain = function(target, effect)
+    if
+        target:getMod(xi.mod.PELICAN_RING_EFFECT) < 2
+    then
```
`if target:getMod(xi.mod.PELICAN_RING_EFFECT) < 2 then`

paladindamarus replied:
I'm okay to change that if you want, but the style guide on the wiki says to avoid single-line conditions:
![image](https://github.com/LandSandBoat/server/assets/105614059/30f30ad9-4388-4dc0-9241-09ee9f0e1857)

Good either way, just want to confirm!

zach2good replied:
```lua
    if target:getMod(xi.mod.PELICAN_RING_EFFECT) < 2 then -- This is a single condition
        target:addMod(xi.mod.PELICAN_RING_EFFECT, 1)
    end
```

Is what Xaver was getting at.

```lua
if singleCondition then
   ....
end

if
    conditionA and
    condition
then
    ...
end
```

paladindamarus replied:
Just realized I totally misunderstood what you were getting at here -- sorry about that!  I'll adjust this one.

Xaver-DaRed commented:
I would recomend:
- Make the changes requested by CI
- Forget Pelican ring. Can be done in a different PR. The smaller the PR, the better for everyone involved.
- This will need testing on base LSB

paladindamarus commented:
> I would recomend:
> 
> * Make the changes requested by CI
> * Forget Pelican ring. Can be done in a different PR. The smaller the PR, the better for everyone involved.
> * This will need testing on base LSB

Will do, thanks!  I'll remove the pelican ring from the items, but leave it as the reward (it will just be a useless item until fixed).  
As for testing:  I didn't state it, but the branch I built this from (despite coming from an ASB repo) is a pure LSB branch pulled from lsb/base.  Only done that way because I can't create another fork from both LSB and ASB at the same time.  However, this was done on my pure LSB installation.

zach2good commented:
Leaving a comment so I get email updates. I'll be doing the C++ reviews once CI passes

zach2good commented on scripts/zones/Selbina/Zone.lua:
```diff
@@ -7,12 +7,19 @@ local zoneObject = {}
     SetServerVariable('Selbina_Destination', math.random(1, 100))
 end
 
+zoneObject.onZoneTick = function(zone)
```
If there are no players in the zone, the contest won't tick - is this intentional?

paladindamarus replied:
Yes -- The tick is only checking to see if the stage progression should trigger (and then progressing if it should).  Once someone is in the zone, it will progress.

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+-- Add history for player fishing data
+-- --------------------------------------------------------
+
+CREATE TABLE IF NOT EXISTS `char_fishing` (
```
If this is just the history for a given player, then this should be `char_fishing_history`

Does this also need to be added to `triggers.sql` so it isn't left behind on char deletion? Or is it more like AH history?

paladindamarus replied:
Good point on triggers -- I did not add it there, but will.  

This is an on-going history of the player's fishing, as there is an NPC they can talk to (to be added in separate PR) that will tell them how many fish they've ever caught, biggest fish, etc.

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+-- --------------------------------------------------------
+-- Add history for player fishing data
```
This comment isn't like the other ones you've added, make it more like those - or remove it. It's pretty obvious these days that all of these files are defining tables.

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+-- Add history for player fishing data
+-- --------------------------------------------------------
+
+CREATE TABLE IF NOT EXISTS `char_fishing` (
```
All the other definitions are DROP then CREATE

paladindamarus replied:
Would that not force the table to be dropped anytime someone ran the file even accidentally?  I thought that was only used for static data?  Does it just get added to the "protected" list in dbtool then?

cocosolos replied:
Add them to the protected table in dbtool and it'll handle making sure there's no overwrites. I'm actually curious _why_ we were dropping the tables in the first place and not just using `IF NOT EXISTS` like here. It doesn't even seem to be consistent across all protected tables. Either way, adding it to dbtool helps with the "lite backup" option.

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+    -- Fishing Catch History
+    `charid`           int(10) unsigned NOT NULL,
+    `fish_list`        blob,
+    `fish_linescast`   int(6)  unsigned NOT NULL DEFAULT 0,
+    `fish_reeled`      int(5)  unsigned NOT NULL DEFAULT 0,
+    `fish_longest`     int(4)  unsigned NOT NULL DEFAULT 0,
+    `fish_longest_id`  int(4)  unsigned NOT NULL DEFAULT 0,
+    `fish_heaviest`    int(4)  unsigned NOT NULL DEFAULT 0,
+    `fish_heaviest_id` int(4)  unsigned NOT NULL DEFAULT 0,
```
Out of interest, why are the display widths for these all a bit different?

paladindamarus replied:
I feel like I may have typoed some of those.  I intended for them to be roughly in line with what would make sense (would someone actually cast 1 million lines?  Maybe...) but then the IDs should be 5-wide and not 4.  I can adjust these to be standardized, but not sure if there's a preferred way of having them done.

paladindamarus replied:
For info, I've removed these entries in the upcoming push as they will be separate when I add the fishing history

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+    `fish_heaviest_id` int(4)  unsigned NOT NULL DEFAULT 0,
+
+    -- Fishing Contest History
+    `contest_rank_1`   smallint(3)  unsigned NOT NULL DEFAULT 0,
+    `contest_rank_2`   smallint(3)  unsigned NOT NULL DEFAULT 0,
+    `contest_rank_3`   smallint(3)  unsigned NOT NULL DEFAULT 0,
+    `contest_rank_4`   smallint(3)  unsigned NOT NULL DEFAULT 0,
```
Why are catch and contest history split here, but are part of the same table? Is this hinting that they should be separate tables?

paladindamarus replied:
No reason they can't be.  I only grouped them because they're both history of the character's fishing activities.  They can be char_fishing_history and char_fishing_contest_history if that's preferred.

zach2good commented on sql/fishing_contest.sql:
```diff
@@ -0,0 +1,13 @@
+-- This table holds the parameters for the current fish ranking contest
+--
+
+CREATE TABLE IF NOT EXISTS `fishing_contest` (
```
Is this going to be a single entry at all times, or does the entire history forever need to be retained (I have no idea how fishing or contests work)

paladindamarus replied:
For now, it's a single entry.  I built the original of this to capture all history, but took that out since it's not really something that retail does.  With this method, there should only be a single entry in the table, ever.  I could do this as a server var as well, but this made it easier to have the data not buried by everything else.

zach2good replied:
I think it's fine to have the table with a single row, we do something very similar with conquest I think, and will probably do the same in the future with besieged - no need to stuff everything into server vars

zach2good commented on sql/fishing_contest_entries.sql:
```diff
@@ -0,0 +1,21 @@
+-- This table holds the submissions to the fish ranking contest
+--
+
+CREATE TABLE IF NOT EXISTS `fishing_contest_entries` (
+    `name`        varchar(15)             NOT NULL,
+    `mjob`        tinyint(2)  unsigned    NOT NULL,
+    `sjob`        tinyint(2)  unsigned    NOT NULL,
+    `mlevel`      tinyint(2)  unsigned    NOT NULL,
+    `slevel`      tinyint(2)  unsigned    NOT NULL,
+    `race`        tinyint(2)  unsigned    NOT NULL,
+    `allegiance`  tinyint(2)  unsigned    NOT NULL,
+    `fishrank`    tinyint(2)  unsigned    NOT NULL,
+    `score`       int(4)      unsigned    NOT NULL DEFAULT 0,
+    `submittime`  int(4)      unsigned    NOT NULL DEFAULT 0,
+    `contestrank` tinyint(2)  unsigned    NOT NULL DEFAULT 0,
+    `share`       tinyint(2)  unsigned    NOT NULL DEFAULT 0,
+    `claimed`     tinyint(1)  unsigned    NOT NULL DEFAULT 0,
+    PRIMARY KEY (`name`) -- Only one entry per character
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```
Is there a reason why a lot of this data isn't pulled from the submitting char with a JOIN?

cocosolos replied:
Also `name` isn't necessarily unique. IDK if that could lead to any issues, but probably want to use charid instead and a join like Zach said.

paladindamarus replied:
Only done this way for consistency with the transmitted packet.  Can very easily be a join.  However, if a character gets deleted, their contest entry should remain intact on the leaderboard, by name.

cocosolos replied:
The issue I'm seeing is that if a character is deleted it'll still exist in the chars table, and a new char could be created with the same name. Would that prevent the new character from also entering the contest?

paladindamarus replied:
> The issue I'm seeing is that if a character is deleted it'll still exist in the chars table, and a new char could be created with the same name. Would that prevent the new character from also entering the contest?

In theory yes, but that's a case I'm not sure how retail handles.  I could simply remove the entry if the character is deleted.

paladindamarus replied:
As I just played around with this, I think it's far more straightforward to just use the name in the table rather than the charid.  If I use the ID, I need to do another table lookup any time it gets references as the player is not guaranteed to be online if another player pulls up the list, for example.  I'd be more worried about lookup errors using the ID.

cocosolos replied:
Not sure about retail either, and it's definitely a fringe issue. Plus I was made aware that we don't actually even open a name back up for reuse on deletion yet, but should in the future. I would use charid as primary key since that for sure will never be changed or reused under normal circumstances and just pull all of the char data that is currently here with a join instead, including name.

cocosolos replied:
> If I use the ID, I need to do another table lookup any time it gets references as the player is not guaranteed to be online if another player pulls up the list, for example. I'd be more worried about lookup errors using the ID.

Not sure what you mean. Player being online or not doesn't change anything about char lookup by charid.

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+
+void CLuaBaseEntity::submitContestFish(uint32 score)
+{
+    if (auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
```
+1 for correct use of dynamic_cast, probably also want a warning for the cast where this fails (in all bindings too)

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+                            "WHERE  name = '%s' "
+                            "AND    claimed != 1; ";
+
+        int32 ret = _sql->Query(Query, PChar->getName());
```
Please use the new `db::` system with `preparedStatement()`, there's lots of examples in `xi_world` and `xi_search`

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+        }
+    }
+
+    return reward;
```
Is an empty table being returned herer OK?

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+{
+    auto table = lua.create_table();
+
+    if (m_PBaseEntity->objtype == TYPE_PC)
+    {
+        auto history = fishingcontest::GetFishingContestHistory(static_cast<CCharEntity*>(m_PBaseEntity));
```
Use dynamic cast inside the if here and then pass to GetFishingContestHistory

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+    if (m_PBaseEntity->objtype == TYPE_PC)
+    {
+        auto history = fishingcontest::GetFishingContestHistory(static_cast<CCharEntity*>(m_PBaseEntity));
+        table[1]     = get<0>(history);
+        table[2]     = get<1>(history);
+        table[3]     = get<2>(history);
+        table[4]     = get<3>(history);
```
Could these table entries be named and then not extracted by index?

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+        table[4]     = get<3>(history);
+    }
+
+    return table;
```
Is an empty table here OK?

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -17709,6 +17710,98 @@ uint8 CLuaBaseEntity::getMannequinPose(uint16 itemID)
+{
+    if (auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
+    {
+        return fishingcontest::ClaimContestReward(PChar);
```
function is marked as returning void, but there is a return here - which is intended?

paladindamarus replied:
Ahhh whoops.  It should be void, which is why it worked and I didn't pick it up.  Will adjust.

zach2good commented on src/map/lua/lua_item.cpp:
```diff
@@ -370,6 +370,31 @@ void CLuaItem::setExData(sol::table const& newData)
+        return sol::lua_nil;
+    }
+
+    if (auto PFish = static_cast<CItemFish*>(m_PLuaItem))
```
Using static_cast here means that ANY non nullptr pointer will pass this check and garbage will go inside the if statement

paladindamarus replied:
This may disappear when I push the updates, but I've removed this function completely since I've realized that I can't cast to a fish item.  Instead, I'm just using the getExData function and working with the data in lua directly instead.

WinterSolstice8 replied:
> I can't cast to a fish item. 

Would this work?

https://github.com/LandSandBoat/server/pull/5704/files

Edit: this has a `new` in it so be careful about the life of the object

paladindamarus replied:
Winter:  I just saw that now.  It probably _would_ work, but I've removed any casting to a `CFishItem` right out of the code and just use the ExData functions.  When I wrote this on Horizon, we didn't have the `getExData` and `setExData`, but those work perfectly for this.

zach2good commented on src/map/lua/lua_item.cpp:
```diff
@@ -370,6 +370,31 @@ void CLuaItem::setExData(sol::table const& newData)
+
+        return table;
+    }
+    else
```
This else is redundant, you'll get here if the if fails because of the return (we still want the logging and the lua_nil return)

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -5724,4 +5732,47 @@ namespace luautils
+    {
+        // Note: Validation of fish item id should be done in lua, since the fish table is also in lua
+        fishingcontest::SetContestFish(fishId);
+        fishingcontest::SetContestMeasure((FISHING_CONTEST_MEASURE)std::min((uint8)measure, (uint8)CONTEST_MEASURE_SMALLEST));
+        fishingcontest::SetContestCriteria((FISHING_CONTEST_CRITERIA)std::min((uint8)criteria, (uint8)CONTEST_CRITERIA_BOTH));
```
Don't use c-style casting, especially not if you have enums. Make your function take the enums as their args

paladindamarus replied:
Not sure I understand this one -- the `std::min` function wont take the enums as they are, but spits out an `int`, hence the casting to the enum to pass into the function.

zach2good replied:
Don't worry about it for now, this'll need extra plumbing in core for the safety I eventually want. Just make sure you replace your c-style casts with static_casts etc.

(the plumbing I'm looking at: https://github.com/Neargye/magic_enum)

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -5724,4 +5732,47 @@ namespace luautils
+
+    void ProgressFishingContest()
+    {
+        // This function will get called every tick when automatic progression is enabled
```
Even if fishing is turned off?

paladindamarus replied:
Not if the fishing contest is turned off... but, I could easily modify the lua calls to only reach into here if both fishing contest is automatic AND fishing is enabled, if that works?

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -5724,4 +5732,47 @@ namespace luautils
+
+    void InitializeFishingContestSystem()
+    {
+        // IMPORTANT: This should only be called on the Zone Init in Selbina
```
Could you add safety checks for this (and logging) to make sure this is enforced?

zach2good commented on src/map/packet_guard.cpp:
```diff
@@ -35,6 +35,7 @@ namespace PacketGuard
         allowList[SUBSTATE_IN_CS][0x016] = true; // Entity Information Request
         allowList[SUBSTATE_IN_CS][0x01A] = true; // Player Action
         allowList[SUBSTATE_IN_CS][0x03A] = true; // Sort Inventory
+        allowList[SUBSTATE_IN_CS][0x04B] = true; // Fish Ranking Packet
```
👍

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+        // The Message Chunk acts as a "sub-type" for the request
+        // 1 = First packet of ranking table
+        // 2 = Subsequent packet of ranking table
+        // 10 = ???
+        // 11 = ??? Prepare to withdraw?
```
Are these unknown, or the actual messages (I don't know fishing at all)

paladindamarus replied:
These are unknown.  I saw them in the packets, but there was no discernable difference with those msg_chunk values.

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+        // 13 = Fish Rank Self, including the score and rank (???) following fish submission (How is it ranked??)
+        std::vector<FishingContestEntry*> entries;
+
+        uint8                maxFakes    = 15;
+        uint8                maxEntries  = 100;
```
Can/should these be broken out into settings?

paladindamarus replied:
Yes and no -- The 100 max limit is a limit on the client (it wouldn't handle more than 100).  The 15 fake entries is what shows on retail, but it will work with or without them.  That can be an easy option.

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+        uint8                maxFakes    = 15;
+        uint8                maxEntries  = 100;
+        uint8                realEntries = fishingcontest::FishingRankEntryCount();
+        uint8                fakeEntries = realEntries >= maxFakes ? 0 : (uint8)std::min((int)maxFakes, maxFakes - realEntries); // Max 15 Fakes
```
No c-style casts

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+        uint8                maxEntries  = 100;
+        uint8                realEntries = fishingcontest::FishingRankEntryCount();
+        uint8                fakeEntries = realEntries >= maxFakes ? 0 : (uint8)std::min((int)maxFakes, maxFakes - realEntries); // Max 15 Fakes
+        uint8                msg_entries = std::min(maxEntries, (uint8)(realEntries + fakeEntries));                             // Max 100 entries
```
No c-style casts

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+        uint8                msg_entries = std::min(maxEntries, (uint8)(realEntries + fakeEntries));                             // Max 100 entries
+        FishingContestEntry* entry       = nullptr;
+        uint8                entryVal    = 0;
+        uint8                blockSize   = sizeof(FishingContestEntry); // Should be 36
```
Can you enforce this with a static_assert somewhere?

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+                else
+                {
+                    FishingContestEntry* original = entry;
+                    entry                         = new FishingContestEntry;
```
I'm _very_ surprised CI hasn't gotten angry here: no use of raw new/delete. Use shared/unique_ptr if you need heap allocation

cocosolos replied:
Looks like it checks for `delete` but not `new`.

paladindamarus replied:
TIL.  I'll work that in!

zach2good replied:
Ah, I remember now, we still use a lot of raw `new` in the packet system and I haven't had time to fix it yet. The point is that if you `new` up memory, you're going to leak it if you don't eventually and consistently `delete` it later on. Smart pointers handle this for you when they fall out of scope (or their ref count reaches zero). 

It's important here because it's in a packet handler, so a sufficiently motivated bad actor could send in many bad packets, make the server allocate a bunch of memory that'll never be free'd, and then starve the process/machine of resources

zach2good replied:
and as a general rule about C++, initialisation is scary and you should be careful! I'll try and explain as best I can:

```cpp
// default init
// default constructor will be called, but member vars won't be default initialised (despite the name), and so will be left uninitialised - typically you'd never use this. Uninitialised memory is garbage, remember.
// You've manually assigned all the members of FishingContestEntry a default value, so those will be used, but this isn't a typical use-case.
auto a = new Object;

// value init
// default constructor will be called, and member vars will be value initialised depending on their types - which is much safer.
auto a = new Object();
```

zach2good replied:
Behold, nightmares: https://en.cppreference.com/w/cpp/language/initialization

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+                    FishingContestEntry* original = entry;
+                    entry                         = new FishingContestEntry;
+
+                    strcpy(entry->name, PChar->name.c_str());
```
Can this be strcpy_n, or whatever the bounded version is? Also, std::

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+            }
+            else // Get the base data if there's no player data
+            {
+                entry = new FishingContestEntry;
```
As above, no raw new

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -8424,7 +8529,7 @@ void PacketParserInitialize()
     PacketSize[0x041] = 0x00; PacketParser[0x041] = &SmallPacket0x041;
     PacketSize[0x042] = 0x00; PacketParser[0x042] = &SmallPacket0x042;
-    PacketSize[0x04B] = 0x0C; PacketParser[0x04B] = &SmallPacket0x04B;
+    PacketSize[0x04B] = 0x00; PacketParser[0x04B] = &SmallPacket0x04B;
```
Does this HAVE to be 0x00 (unchecked) because the size is variable, or can you look up what the size is meant to be?

paladindamarus replied:
It's because the size is variable depending on the type.  Packet 0x04B is size 0x0C for a regular server message packet, but different size for the ranking packet (I don't remember the actual size ... 0x18 I think?)  I can re-investigate this.

zach2good commented on src/map/packets/fishing_rank.cpp:
```diff
@@ -0,0 +1,84 @@
+﻿/*
+===========================================================================
+
+  Copyright (c) 2010-2015 Darkstar Dev Teams
```
Wrong year/project name

zach2good commented on src/map/packets/fishing_rank.h:
```diff
@@ -0,0 +1,38 @@
+﻿/*
+===========================================================================
+
+  Copyright (c) 2010-2015 Darkstar Dev Teams
```
Same here

zach2good commented on src/map/packets/fishing_rank.h:
```diff
@@ -0,0 +1,38 @@
+===========================================================================
+*/
+
+#ifndef _CFishRankingPacket_H
```
I dont remember what newer files use, please check to see if we use pragma once or header guards - if we use header guards they're all caps though

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+         *  See enum FISHING_CONTEST_STATUS in fishingutils.h
+         *  CALLED BY ZONE TICK IN SELBINA - DO NOT ADD TO FUNCTION
+         ***********************************************/
+        uint32 currentTime = (uint32)time(NULL);
```
use C++ time functions

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+    {
+        // This should only be run once per contest, once the contest progresses from "Accepting" to "Releasing"
+        // Traverse the entries and update the database for each player
+        for (FishingContestEntry* it : FishingContestEntries)
```
rename it to entry, it isn't an iterator in this case

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+            {
+                // Get the rank for column header
+                uint8 rankGroup = 0;
+                if (it->contestRank > 0 and it->contestRank <= 4)
```
we don't use verbal `and` or `or` (I don't know how CI hasn't caught all of this...), use the ampersand and pipe versions

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+        TracyZoneScoped;
+
+        // Sort the list first
+        // clang-format off
```
Good try! you've wrapped the lambda correctly, but the formatting inside is wrong 😆 

It's truly a shame that clang-format hasn't made this possible yet

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+        }
+    }
+
+    auto GetFishingContestHistory(CCharEntity* PChar) -> std::tuple<uint16, uint16, uint16, uint16>
```
Can this tuple be a named struct? This is very anonymous

paladindamarus replied:
absolutely -- I only did it this way because it's only used once to return the rank 1 / rank 2 / rank 3 / rank 4-10 values.  I'll build a struct!

zach2good commented on src/map/utils/fishingcontest.h:
```diff
@@ -0,0 +1,150 @@
+{
+    CONTEST_CRITERIA_SIZE   = 0,
+    CONTEST_CRITERIA_WEIGHT = 1,
+    CONTEST_CRITERIA_BOTH   = 2
```
Missing ending commas on some of these enums

zach2good commented on src/map/utils/fishingcontest.h:
```diff
@@ -0,0 +1,150 @@
+class CBasicPacket;
+class CCharEntity;
+
+enum FISHING_CONTEST_CRITERIA : uint8
```
Can these be enum classes?

zach2good commented on src/map/utils/fishingcontest.h:
```diff
@@ -0,0 +1,150 @@
+    FISHING_CONTEST_CRITERIA criteria   = CONTEST_CRITERIA_SIZE;
+    FISHING_CONTEST_MEASURE  measure    = CONTEST_MEASURE_GREATEST;
+    uint16                   fishId     = 0;
+    uint32                   startTime  = 0xEFFFFFFF; // to allow for changetimes to not overflow
```
Can you explain this? It isn't immediately obvious how this works

paladindamarus replied:
I can probably get rid of this and just use 0 as the start time -- When I initially coded this on Horizon, I was storing completed contests in the database and starting new ones in cache before updating the DB, so I needed a start time trigger of '0' to flag the contest as beginning, and I used a guaranteed future time just to show that it's in a creation state.  I didn't use 0xFFFFFFFF because if it somehow got through and calculated stage timings, it would overflow.

zach2good reviewed (changes requested):
Only glanced though the C++ changes, it looks like the C++ CI hasn't run properly... (or we're missing rules)

zach2good commented:
See this section in dbtool.py

# These are the 'protected' files
player_data = [
    "accounts.sql",
    "accounts_banned.sql",
    "auction_house_items.sql",
    "auction_house.sql",
    "char_blacklist.sql",
    "char_chocobos.sql",
    "char_effects.sql",
    "char_equip.sql",
...

On Tue, Jul 9, 2024 at 10:31 PM Damarus ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sql/char_fishing.sql
> <https://github.com/LandSandBoat/server/pull/5990#discussion_r1671264041>:
>
> > @@ -0,0 +1,22 @@
> +-- --------------------------------------------------------
> +-- Add history for player fishing data
> +-- --------------------------------------------------------
> +
> +CREATE TABLE IF NOT EXISTS `char_fishing` (
>
> Would that not force the table to be dropped anytime someone ran the file
> even accidentally? I thought that was only used for static data? Does it
> just get added to the "protected" list in dbtool then?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LandSandBoat/server/pull/5990#discussion_r1671264041>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKTJIPLKJLONAPSHY2ABUTZLRJEFAVCNFSM6AAAAABKTNZTX6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNRXGUZTQMZTGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

zach2good commented on scripts/commands/addfish.lua:
```diff
@@ -0,0 +1,88 @@
+        return
+    end
+
+    -- Give the GM the item...
+    local obtained = player:addItem({ id = itemToGet,
+                                        quantity = 1,
+                                        exdata =
+                                        {
+                                            [0] = bit.band(length, 0x00FF),
+                                            [1] = bit.rshift(bit.band(length, 0xFF00), 8),
+                                            [2] = bit.band(weight, 0x00FF),
+                                            [3] = bit.rshift(bit.band(weight, 0xFF00), 8),
+                                            [4] = ranked,
+                                        } })
+    if obtained then
+        player:messageSpecial(ID.text.ITEM_OBTAINED, itemToGet)
+    end
```
Can/should this be done with npcUtils?

paladindamarus replied:
I would like it to be but I'd need to add code to npc_util to accept exdata.  I can certainly do that if you want me to build it into this PR.

zach2good commented on scripts/commands/addfish.lua:
```diff
@@ -0,0 +1,88 @@
+        return
+    end
+
+    -- TODO: check qty and stack size + remaining inventory space instead of hardcoded == 0 check
```
Address TODO

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+
+xi.fishingcontest.fish =
+{
+    -- NOTE: If you don't want certain fish included in the contest (era restrictions, or whatever),
```
Good note 👍

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+require('scripts/globals/utils')
+-----------------------------------
+xi = xi or {}
+xi.fishingcontest = xi.fishingcontest or {}
```
xi.fishingContest, like others: `xi.additionalEffect = xi.additionalEffect or {}`

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+-----------------------------------
+-- Data for Fishing
```
If this is _just_ the data, and there's a driver somewhere else the whole file can/should be `fishing_constest_data.lua`

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+    return 0
+end
+
+-- The npcUtil.giveItem function currently does not support custom exdata
```
Would be useful to add exdata to npcUtil.giveItem?

paladindamarus replied:
I'm happy to do that if you want it in this PR, but yes -- it would definitely be useful (probably for others too, I assume?)

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+    local fishid = params.id
+    -- does player have enough inventory space?
+    if player:getFreeSlotsCount() < 1 then
+        local messageId = (ID.text.ITEM_CANNOT_BE_OBTAINED + 4) or ID.text.ITEM_CANNOT_BE_OBTAINED
```
I don't understand what's happening here? In Lua the only things that are falsy are nil and false, not 0. So if any part of ID.text.ITEM_CANNOT_BE_OBTAINED is nil, then so will the other or side

paladindamarus replied:
I copied this from the `npcUtil.giveItem` function to be consistent as an interim (as mentioned before, the npcUtil functions don't currently handle exdata so I had to make my own for now):
![image](https://github.com/user-attachments/assets/89d192a1-158e-4787-bb49-9a06b08021dc)

I can adjust this, but I'm not really sure what to adjust it to if the goal is to guarantee that the ITEM_CANNOT_BE_OBTAINED is not null (or more specifically, equals a particular value)

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+-----------------------------------
+xi.fishingcontest.selectContestFish = function()
+    local fishList = xi.fishingcontest.fish
+    return xi.fishingcontest.fish[math.random(#fishList)]['id']
```
We have utils to wrap this so you can do utils.randomEntry(fishList)['id'], I think

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+end
+
+xi.fishingcontest.onStageProgress = function(newStage)
+    -- Place holder -- I believe there is an announcement made by the NPC at certain stage progressions
```
TODO this

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+xi.fishingcontest.onTrigger = function(player, npc)
+    local contest = GetFishingContest()
+
+    player:startEvent(10006, {
```
Sneaky sneaky, I totally forgot startEvent() could take a table in it's sol::variadic_args

zach2good commented on scripts/globals/fishing_contest.lua:
```diff
@@ -0,0 +1,417 @@
+    if csid == 10007 then
+        -- Give the player back a fish with the same stats, but tagged for Ranking
+        if option == 145 then
+            local length = player:getLocalVar('[FishContest]Length')
```
Is this needed on either side of the trade? Or is this a thing that can be lost if the player zones?

paladindamarus replied:
There shouldn't be any zoning since this is during the interaction with the NPC.  The local var is just a temporary holder for the fish stats so that they can receive the same fish back but marked for ranking (i.e. so they cannot use it a second time).

zach2good commented on sql/char_fishing.sql:
```diff
@@ -0,0 +1,22 @@
+-- Add history for player fishing data
+-- --------------------------------------------------------
+
+CREATE TABLE IF NOT EXISTS `char_fishing` (
```
As I'm looking at this more (and I still don't know how any part of fishing works), this is denormalised right? We'll only ever have one row per char which we read, update, and then write back to the db with their best results etc.? 

We absolutely cannot have a forever history of  every single entry of every fish of every char that has been submitted to every contest etc. - because we'll then have to search over it and having that historical data isn't worth the cost (like it is for AH, which we have utils to truncate)

paladindamarus replied:
That is correct.  At the start of every contest, the entry table is wiped clean, and the only thing that gets saved is the number of times a player has achieved either rank 1, rank 2, rank 3, or rank 4-10.  One row per char only.,

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -230,6 +231,13 @@ namespace luautils
+        lua.set_function("InitNewFishingContest", &luautils::InitNewFishingContest);
+        lua.set_function("SetContestParameters", &luautils::SetContestParameters);
+        lua.set_function("ProgressFishingContest", &luautils::ProgressFishingContest);
+        lua.set_function("InitializeFishingContestSystem", &luautils::InitializeFishingContestSystem);
```
I _think_ you'll have to add these to the Lua ci scripts if you haven't already, so it doesn't complain about unknown global funcs

paladindamarus replied:
Is that `global_objects` in lua.sh?

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,133 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
-    const char* fmtQuery = "SELECT version_mismatch FROM accounts_sessions WHERE charid = %u";
-    int32       ret      = _sql->Query(fmtQuery, PChar->id);
-    if (ret != SQL_ERROR && _sql->NextRow() == SQL_SUCCESS)
+    // uint8  msg_unknown1   = data.ref<uint8>(0x05);  // Unknown always 0
```
Could you check with atom0s and see if he knows what the other unknown parts of this packet are? He's usually great for this sort of thing

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+        }
+
+        // Clear the fishing contest entries from cache and database
+        FishingContestEntries.clear();
```
In the current form, this is a memory leak of all the fishing contest entries - the vector of pointers is cleared, but the memory those pointers point to is not.

You're going to refactor things to be smart pointers, so this code won't need to change - it'll work as-is. 

If you were to keep the current manual memory management this would need to look like this:
```cpp
for (auto entry : FishingContestEntries)
{
    delete entry;
}
FishingContestEntries.clear()
```

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,716 @@
+
+namespace fishingcontest
+{
+    FishingContest                    CurrentFishingContest; // Currently active contest
+    std::vector<FishingContestEntry*> FishingContestEntries; // Currently active contest entries
```
If these are only used in here, put them into an anonymous namespace (I think you can see this in zone_entities)

zach2good reviewed (commented):
Another round of reviews before I start work

zach2good commented:
And as a general question: did this work start off somewhere else and does it need git attribution/co-author data, or is it entirely new? I say that because of the DSP headers on the C++ files, and some familiar-looking patterns in the C++ code

paladindamarus commented:
> And as a general question: did this work start off somewhere else and does it need git attribution/co-author data, or is it entirely new? I say that because of the DSP headers on the C++ files, and some familiar-looking patterns in the C++ code

This is entirely my work alone, except for some data from captures (some mine, some from the capture discord), and the packet header data that I got from the PV tool / atom0s.  I wrote it initially specifically on Horizon, but it had a bunch of extra features that I'm ripping out (hence why are there a few of the oddities in the code -- i refactored and missed some things).  The DSP headers were just me copying a new file header to start, and I never went back to change the header.  I probably copied some patterns from other spots in the code, mostly looking for the proper structure of things.

paladindamarus commented:
Latest batch of updates applied, addressing comments above.  Note that the one check that failed seems to be a socket binding error that I cannot fix on my end.  Everything else seems good.

zach2good commented on src/map/utils/fishingutils.cpp:
```diff
@@ -2891,8 +2891,11 @@ namespace fishingutils
+            if (CItemUsable* PItemUsable = dynamic_cast<CItemUsable*>(PItem))
+            {
+                // CItemFish constructor uses `const CItem&` input so this is ok
+                return new CItemFish(*PItemUsable);
```
Even if this was here before you got here, it's still a memory leak. You can fix it like so: https://godbolt.org/z/dKoez1hvc

paladindamarus replied:
This is way out of my scope here.  I don't mind making this fix, but this touches the entire fishing system.  I've reset this back to the exact way it is on LSB now (will be updated in next push).  

In looking at this though ... I've realized that `GetFish` is overloaded in fishingutils.cpp:

`fish_t*                   GetFish(uint32 fishId);`
`CItemFish*       GetFish(uint16 itemid); // creates a 'new' CItemFish if possible`

WinterSolstice8 replied:
Doesn't leak if any CItem or derivative makes it into a container:
```
CItemContainer::~CItemContainer()
{
    for (uint8 SlotID = 0; SlotID <= m_size; ++SlotID)
    {
        destroy(m_ItemList[SlotID]);
    }
}
```

zach2good replied:
I'll handle this is some ongoing work I have: https://github.com/LandSandBoat/server/pull/5185

zach2good replied:
I suppose it's out of scope for this PR, but here is the usage:
```cpp
    int32 CatchFish(CCharEntity* PChar, uint16 FishID, bool BigFish, uint16 length, uint16 weight, uint8 Count = 1)
    {
        uint16 MessageOffset = GetMessageOffset(PChar->getZone());
        PChar->animation     = ANIMATION_FISHING_CAUGHT;
        PChar->updatemask |= UPDATE_HP;

        if (PChar->getStorage(LOC_INVENTORY)->GetFreeSlotsCount() != 0)
        {
            CItemFish* Fish = GetFish(FishID);

            if (Fish == nullptr)
            {
                ShowError("Invalid ItemID %i for fished item\n", FishID);
                PChar->animation = ANIMATION_FISHING_STOP;
                PChar->updatemask |= UPDATE_HP;
                PChar->pushPacket(new CMessageTextPacket(PChar, MessageOffset + FISHMESSAGEOFFSET_LOST));
                return 0;
            }

            if (BigFish && length > 1 && weight > 1)
            {
                Fish->SetLength(length);
                Fish->SetWeight(weight);
            }

            Fish->setQuantity(Count);
            charutils::AddItem(PChar, LOC_INVENTORY, Fish);

            if (Count > 1)
            {
                PChar->loc.zone->PushPacket(PChar, CHAR_INRANGE_SELF, new CCaughtFishPacket(PChar, FishID, MessageOffset + FISHMESSAGEOFFSET_CATCH_MULTI, Count));
            }
            else
            {
                PChar->loc.zone->PushPacket(PChar, CHAR_INRANGE_SELF, new CCaughtFishPacket(PChar, FishID, MessageOffset + FISHMESSAGEOFFSET_CATCH, Count));
            }

            return 1;
        }
        else
        {
            PChar->loc.zone->PushPacket(PChar, CHAR_INRANGE_SELF, new CCaughtFishPacket(PChar, FishID, MessageOffset + FISHMESSAGEOFFSET_CATCH_INV_FULL, Count));
        }

        return 0;
    }
```

Quite right, if it ends up in a container it'll be cleaned up eventually, but there's a path here that contains a leak. I'll handle this in my other PR when I get around to it.

zach2good commented on src/map/utils/fishingcontest.h:
```diff
@@ -0,0 +1,164 @@
+    void SetContestStartTime(uint32 startTime);
+
+    // Contest Data: Gets
+    FISHING_CONTEST_STATUS   GetContestStatus();
```
With all of these long types, you can use auto and trailing return syntax to make it less eye-bleedy

zach2good commented on src/map/utils/fishingcontest.h:
```diff
@@ -0,0 +1,164 @@
+    CONTEST_MEASURE_SMALLEST = 1,
+};
+
+enum class FISHING_CONTEST_STATUS : uint8
```
Notice that the vast majority of all our other enum class declarations don't repeat the name of the enum:
```cpp
enum class Emote : uint8
{
    POINT      = 0,
    BOW        = 1,
    SALUTE     = 2,
    KNEEL      = 3,
    LAUGH      = 4,
    CRY        = 5,
    NO         = 6,
    YES        = 7,
    ...
```

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,790 @@
+    {
+        // load the overhead contest parameters
+        std::string Query = "SELECT "
+                            "status, "       // 0
```
No need to annotate with the column number, we look up by name as you've already done below

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,790 @@
+        }
+
+        // Clamp vars, just in case
+        criteria = (FISHING_CONTEST_CRITERIA)std::min((uint8)criteria, (uint8)FISHING_CONTEST_CRITERIA::CONTEST_CRITERIA_BOTH);
```
c-style casts -> static_casts

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,790 @@
+    std::vector<FishingContestEntry> FakeContestEntries;    // Pre-established vector of placeholder entries
+
+    // Define the "next stage" for each stage for automatic progression
+    std::unordered_map<FISHING_CONTEST_STATUS, FISHING_CONTEST_STATUS> nextStage = {
```
This is just a ring, so just define as a vector and access with currentStage  = currentStage % vec.size() so it wraps around, instead of you having to define this order. And then have separate logic for bailing out early if closed.

zach2good commented on src/map/utils/fishingcontest.cpp:
```diff
@@ -0,0 +1,790 @@
+﻿/*
+===========================================================================
+
+Copyright (c) 2024 LandSandBoat Dev Teams
```
This is not formatted like our other headers

paladindamarus replied:
Odd... Can you point me to one that is correct?  I tried to copy one from something recent like the campaign_system.cpp (2023) file and just change the date to 2024.  I'm not sure which one should be the latest.

zach2good replied:
Use regex and count.

```
^This program is free software: you can redistribute it and/or modify
^  This program is free software: you can redistribute it and/or modify
^    This program is free software: you can redistribute it and/or modify
```

zach2good replied:
Clarified in this PR: https://github.com/LandSandBoat/server/pull/6172

paladindamarus replied:
Ah, got it!  Thanks!

zach2good commented on src/map/packets/fish_ranking.cpp:
```diff
@@ -0,0 +1,87 @@
+
+#include <cstring>
+
+CFishRankingPacket::CFishRankingPacket(std::vector<FishingContestEntry>* entries, int8 language, int32 timestamp, int32 message_offset, uint32 totalEntries, uint8 msg_chunk)
```
Don't pass in a pointer, take vec<Entries> by reference.

zach2good replied:
message_offset -> messageOffset 
msg_chunk -> msgChunk

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -2378,29 +2379,122 @@ void SmallPacket0x042(map_session_data_t* const PSession, CCharEntity* const PCh
+
+        // TODO: kill player til theyre dead and bsod
+        const char* fmtQuery = "SELECT version_mismatch FROM accounts_sessions WHERE charid = %u";
+        int32       ret      = _sql->Query(fmtQuery, PChar->id);
```
Use db:: system

WinterSolstice8 commented on src/map/utils/fishingutils.cpp:
```diff
@@ -2891,8 +2891,11 @@ namespace fishingutils
         {
-            // CItemFish constructor uses `const CItem&` input so this is ok
-            return new CItemFish(*PItem);
+            if (CItemUsable* PItemUsable = dynamic_cast<CItemUsable*>(PItem))
```
What is the purpose of this cast?
 `class CItemEquipment : public CItemUsable` 
Neither of these classes are parents or children of CItemFish
What problem were you trying to solve here?

This is weird because I know (some) fish can be eaten by mithras or with some hidden effect items. Maybe CItemFish should be a child of CItemUsable..?

paladindamarus replied:
Yeah I was trying something with casting and forgot to remove this piece.  I've taken this out and am resetting it back to what is on LSB.  I'm just not even using the GetFish function and instead checking the exdata directly.

WinterSolstice8 replied:
Oh ok, makes sense. I still wonder what's going on with usable fish then... hmm...

zach2good commented on src/map/utils/zoneutils.h:
```diff
@@ -43,17 +43,19 @@ namespace zoneutils
+    CCharEntity*        GetCharFromWorld(uint32 charid, uint16 targid);  // returns pointer to character by id and target id
+    CCharEntity*        GetChar(uint32 id);                              // returns pointer to character by id
+    CCharEntity*        GetCharToUpdate(uint32 primary, uint32 ternary); // returns pointer to preferred char to update for party changes
+    std::vector<uint16> GetZonesOnThisProcess();
```
This can be auto and trailing return so you don't have to move all the other funcitons

zach2good commented on tools/dbtool.py:
```diff
@@ -190,6 +190,7 @@ def load_into_dict(filename, settings):
     "char_equip.sql",
     "char_equip_saved.sql",
     "char_exp.sql",
+    "char_fishing_contest_history",
```
Is this likely to get very large over time, and need truncating?

paladindamarus replied:
it shouldn't -- it's one row per character at most.  This is like `char_history`, but only holds the number of times each character has come in either first, second, third, or fourth-tenth place.  It is _not_ an ongoing log of individual contest results.  It will look something like this:

![image](https://github.com/user-attachments/assets/e0fbd46d-d8c9-4736-b282-c548a7e75586)

zach2good reviewed (approved):
Since fishing is off by default, and this has already gone through one round of reviews (by me at least), we may as well merge it before it rots (I have some heavy packet work that will _ruin_ this PR coming soon, so it's easier for everyone if it just goes in)
